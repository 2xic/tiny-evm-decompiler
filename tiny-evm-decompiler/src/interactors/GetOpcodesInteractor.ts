//import { OpCode } from 'tinyeth/dist/evm/OpCode';
import { Opcodes } from 'tinyeth/dist/';
import { OpCode } from 'tinyeth/dist/evm/OpCode';

export class GetOpcodesInteractor {
  public getOpcodes({
    contract
  }: {
    contract: string
  }): ParsedOpcodes[] {
    let contractBuffer = Buffer.from(contract.replace('0x', ''), 'hex')
    const opcodes: ParsedOpcodes[] = [];
    let address = 0;
    let isReadingCode = true;
    let codeBlockType = CodeBlockType.START;

    while (contractBuffer.length) {
      const currentOpcode = contractBuffer[0];
      const opcode: OpCode | undefined = Opcodes[currentOpcode];

      if (opcode?.mnemonic === 'JUMPDEST'){
        isReadingCode = true;
        codeBlockType = CodeBlockType.START;
      }

      if (!opcode) {
        opcodes.push({
          address,
          opcode: 'INVALID',
          value: currentOpcode,
        });
        contractBuffer = contractBuffer.slice(1);
        address += 1;
      } else if (!isReadingCode){
        opcodes.push({
          address,
          opcode: `DATA 0x${currentOpcode.toString(16)}`,
          value: currentOpcode,
        });
        contractBuffer = contractBuffer.slice(1);
      } else {
        if (opcode.isTerminating){
          isReadingCode = false;
          codeBlockType = CodeBlockType.END
        }
        const opcodeArguments = contractBuffer.slice(1, opcode.length);
        opcodes.push({
          address: address,
          opcode: {
            mnemonic: opcode.mnemonic,
            arguments: Array.from(opcodeArguments).map((item) => `0x${item.toString(16)}`),
          },
          value: currentOpcode,
          codeBlockType,
        });
        contractBuffer = contractBuffer.slice(!opcode.length ? 1 : opcode.length);
        address += opcode.length;
      }

      codeBlockType = undefined;
    }

    return opcodes;
  }
}

export type ParsedOpcodes = {
  address: number,
  opcode: string | {
    mnemonic: string;
    arguments: unknown[];
  };
  value: number;
  codeBlockType?: CodeBlockType,
}

export enum CodeBlockType {
  'START',
  'END'
};
