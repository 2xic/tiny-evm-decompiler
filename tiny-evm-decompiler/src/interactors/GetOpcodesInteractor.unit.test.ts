import { GetOpcodesInteractor } from "./GetOpcodesInteractor"

describe('GetOpcodesInteractor', () => {
  it('should not be fooled by data', () => {
    const interactor = new GetOpcodesInteractor();
    const response = interactor.getOpcodes({
      contract: Buffer.from('600056010203', 'hex')
    })
    expect(response.map((item) => [item.opcode])).toMatchSnapshot()
  })
});
