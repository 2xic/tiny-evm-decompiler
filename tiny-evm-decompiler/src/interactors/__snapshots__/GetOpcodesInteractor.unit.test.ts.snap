// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GetOpcodesInteractor should not be fooled by data 1`] = `
[
  [
    {
      "arguments": "0x00",
      "isReal": true,
      "mnemonic": "PUSH1",
      "opcode": 96,
    },
  ],
  [
    {
      "arguments": "",
      "isReal": true,
      "mnemonic": "JUMP",
      "opcode": 86,
    },
  ],
  [
    {
      "arguments": "0x1",
      "isReal": false,
      "mnemonic": "DATA",
      "opcode": 1,
    },
  ],
  [
    {
      "arguments": "0x2",
      "isReal": false,
      "mnemonic": "DATA",
      "opcode": 2,
    },
  ],
  [
    {
      "arguments": "0x3",
      "isReal": false,
      "mnemonic": "DATA",
      "opcode": 3,
    },
  ],
]
`;
